scipy/linalg/_decomp_cossin.py

```diff
@@ -0,0 +1,186 @@
+    >>> x = unitary_group.rvs(4)
+    >>> u, cs, vdh = cossin(x, (2,2))
+    >>> np.allclose(x, u @ cs @ vdh)
+    True
+
+    References
+    ----------
+    .. [1] : Brian D. Sutton. Computing the complete CS decomposition. Numer.
+           Algorithms, 50(1):33-65, 2009.
+
+    """
+    if partitioning:
+        X = _asarray_validated(X, check_finite=True)
+        if not np.equal(*X.shape):
+            raise ValueError("Cosine Sine decomposition only supports square"
+                             " matrices, got {}".format(X.shape))
+        p, q = partitioning
+        return _cossin(x11=X[:p, :q], x12=X[:p, q:], x21=X[p:, :q],
+                       x22=X[p:, q:])
+
+    if not isinstance(X, Iterable):
+        raise ValueError("When top_left_shape=None, X must be Iterable")
+
+    X = list(X)
+    if len(X) != 4:
+        raise ValueError("When partitioning=None, "
+                         "exactly four submatrices should be in X, got {}"
+                         .format(len(X)))
+
+    x11, x12, x21, x22 = X
+
+    p, q = np.shape(np.atleast_2d(x11))
+    mmp, mmq = np.shape(np.atleast_2d(x22))
+
+    p2, mmq2 = np.shape(np.atleast_2d(x12))
+    mmp2, q2 = np.shape(np.atleast_2d(x21))
+
+    if p != p2 or q != q2 or mmp != mmp2 or mmq != mmq2 or mmq + q != mmp + p:
+        shapes = [np.shape(np.atleast_2d(subx)) for subx in X]
+        raise ValueError("invalid submatrix dimensions {} to build a square "
+                         "matrix, maybe missing partitioning?".format(shapes))
+    return _cossin(x11, x12, x21, x22)
+
+
+def _cossin(x11, x12, x21, x22):
+    p, q = np.shape(x11)
+    m = np.shape(x22)[0] + p
+
+    iscomp = np.iscomplexobj(x11)
+    func_name = "uncsd" if iscomp else "orcsd"
+
+    csd, csd_lwork = get_lapack_funcs([func_name, func_name + "_lwork"],
+                                      [x11])
+    lwork = _compute_lwork(csd_lwork, m=m, p=p, q=q)
+
+    lwork_args = (dict(lwork=lwork[0], lrwork=lwork[1]) if iscomp else
+                  dict(lwork=lwork))
+
+    csd_args = dict(x11=x11,
+                    x12=x12,
+                    x21=x21,
+                    x22=x22,
+                    compute_u1=True,
+                    compute_u2=True,
+                    compute_v1t=True,
+                    compute_v2t=True,
+                    trans=False,
+                    signs=False)
+
+    csd_args.update(lwork_args)
+
+    x11o, x12o, x21o, x22o, theta, u1, u2, v1h, v2h, info = csd(
+        **csd_args)
+
+    method_name = csd.typecode + func_name
+    if info < 0:
+        raise ValueError('illegal value in argument {} of internal {}'
+                         .format(-info, method_name))
+    if info > 0:
+        raise LinAlgError("{} did not converge: {}".format(method_name, info))
+
+    U = block_diag(u1, u2)
+    VDH = block_diag(v1h, v2h)
+    CS = _construct_cs(theta, m, p, q, x11.dtype.type)
+    return U, CS, VDH
+
+
+def _construct_cs(theta, m, p, q, dtype):
```

Comment:
This is a bit too slow with all the for loops, I'll try to cook up something later with less looping.

Reply:
I had [a version that was using blocks](https://github.com/scipy/scipy/blob/7aa3e61d7f472a6e459e990c75887a03059d576e/scipy/linalg/tests/test_decomp_cossin.py#L79-L102) do you think that would work?

Reply:
The blocks version is not faster at all. I find your solution to be efficient. The cosine-sine loop is required, we have at most 2 identity loops. I might be missing something, but unless there is some kind of numpy magic that copies larger arrays more efficiently into a diagonal than just a simple loop, I think this is optimal.